### PR TITLE
object: add ldap configs for rgw

### LINKED ISSUE #4315: Support RGW LDAP integration
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
The Ceph Object Gateway supports integrating with LDAP for authenticating and creating users, see: https://docs.ceph.com/docs/master/radosgw/ldap-auth/
It would be nice if this was directly supported in the "object" part of the rook-ceph integration.
**What is use case behind this feature:**
Providing this feature would enable easy integration with a corporate LDAP server for use of the object store, including its S3 compatible interface.  This would prevent having to have someone create individual user accounts via the "object-user.yaml" for every user that needs access to the object store, and allow for control based on group membership(s) in the LDAP server.
**Environment**:

This can currently be mostly achieved using existing "escape hatches" in rook.  The turning on of the LDAP integration can be done by adding config values to the "rook-config-override" configmap similar to the following (actual values removed):
[global]
    rgw ldap binddn = 
    rgw_ldap_secret = /etc/ceph/ldap/bindpass.secret
    rgw ldap uri = 
    rgw ldap searchdn =
    rgw ldap dnattr =
    rgw_s3_auth_use_ldap = true

The main problem comes from the fact that "rgw_ldap_secret" needs to point to a file that contains the unencrypted/unencoded password for the account being used to query the LDAP server that is specified by "rgw_ldap_binddn".   Mounting this file into the "rook-ceph-rgw-my-store" related pod is difficult/fragile at the moment.  By creating a secret that contains the password, we can mount that secret as a volume in the pod, but that requires modifying/patching the "rook-ceph-rgw-my-store" deployment similar to below:
spec:
  template:
    spec:
      containers:
      - name: rgw
        volumeMounts:
        - mountPath: /etc/ceph/ldap
          name: rook-ceph-rgw-ldap-pass
          readOnly: true
      volumes:
      - name: rook-ceph-rgw-ldap-pass
        secret:
          defaultMode: 420
          secretName: rook-ceph-rgw-ldap-pass

This works, and the LDAP integration succeeds.  But it is fragile, in that anything that causes the operator to restart/redeploy the  "rook-ceph-rgw-my-store" deployment overwrites the above change, and it needs to be reapplied.

### PR DESCRIPTION
**Pull request by jhoblitt**
**Description of your changes:**
Rebase of #8750 on current master + support for `rgw_ldap_searchfilter`.

> The Ceph Object Gateway supports integrating with LDAP for authenticating
> and creating users, for that required config options need to be added in
> cephobjectstore CRD.

**Which issue is resolved by this Pull Request:**
Supersedes #8750
Resolves #4315 

**Checklist:**

- [X] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [X] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [X] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [X] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [X] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by jhoblitt** [comment]
I asked for suggestions in the k8s slack org and there was a suggestion to use apacheds for integration testing: https://github.com/PacktPublishing/Kubernetes---An-Enterprise-Guide-2E/blob/main/chapter5/apacheds.yaml

**Comment by jhoblitt** [comment]
As mentioned in https://github.com/rook/rook/issues/4315, I have gotten an integration test working using an openldap chart.  I believe only some unit tests are required before this work is ready for merge.

**Comment by jhoblitt** [comment]
I've added some basic unit tests for `LDAPSpec`.I think coverage for the ldap secret is still needed.

**Comment by jhoblitt** [comment]
@thotz `LDAPSpec` being `{}` was causing `--rgw-s3-auth-use-ldap` and `--rgw-ldap-secret` to be in the args.  It seems wrong to me to enable ldap without their being an ldap server config.  I added a commit that causes `--rgw-ldap-secret` not to be added to the args unless `CredentialSecretName` is set... but now that I think about it, that is probably wrong too.  Perhaps it should be an error if `LDAPSpec` is empty or if it is, no `--rgw` flags are added to the args?

**Comment by jhoblitt** [comment]
@travisn Would it be possible to find someone for a review of this PR before it rots any further?

**Comment by jhoblitt** [comment]
Rebaed on current master.

**Review by travisn — CHANGES_REQUESTED** [review]
- The new ldap integration test is failing, we will need it to pass consistently before merge.
- Before merge, we'll want most of the commits smashed, and perhaps use the `Co-authored-by` commit tag instead of keeping @thotz's original commit intact.
- @thotz @BlaineEXE As a continuation of #8750, please review.

**Comment by jhoblitt** [comment]
Note that CI was passing prior to the rebase. Several integration tests failed on this run that were unrelated to these changes.

**Comment by travisn** [comment]
I just triggered a re-run to see if it is a consistent failure

**Comment by BlaineEXE** [comment]
@jhoblitt sorry for the delay here on my end. One of the reasons https://github.com/rook/rook/pull/8750 was closed was because it became clear that the feature was sufficiently large enough that it requires a design document. A design document allows us to agree on certain details of the feature before implementation which speeds up review and iteration on the implementation. Our experience in Rook is that it results in implementations that have fewer bugs and that get merged more quickly.

Before we continue here, please submit a PR to add to (or create a new) design doc for CephObjectStore LDAP integration. Things the design doc should answer:
- What is the user API? i.e., what are the fields added to the CephObjectStore resource?
- Can the LDAP be used today with Rook via manual steps? 
  - If so, what are those steps?
  - If not, what is missing in Rook to allow it?
- In order to support LDAP, what will the Rook operator have to configure within Ceph?

You can see these design docs for some idea: https://github.com/rook/rook/blob/master/design/ceph/object/swift-and-keystone-integration.md and https://github.com/rook/rook/blob/master/design/ceph/ceph-nfs-ganesha.md (specifically sssd and kerberos design aspects)

File: deploy/examples/object-openshift.yaml
Comment (travisn) [inline deploy/examples/object-openshift.yaml:106]: I'd suggest let's not update object-openshfit.yaml for the ldap feature. We don't have to replicate all the settings in all the examples.

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:1381]: Is `&()'d` a common term in url construction for appending? For the docs, I might suggest this unless the original is more clear for some people
```suggestion
	// A generic search filter. If `dnattr` is set, this filter is appended
```

**Comment by jhoblitt** [comment]
I have been holding off on updating this PR until https://github.com/rook/rook/pull/11320 is accepted.

**Comment by travisn** [comment]
Reopening since #11320 is almost finalized...

**Comment by jhoblitt** [comment]
I've done a trivial rebase.  I haven't yet gone through to see where this PR is out of sync with the design doc.

**Comment by BlaineEXE** [comment]
I'd like to note a general desire/intent to create an Active Directory (which includes LDAP) test that we could validate LDAP support in NFS and RGW. NFS still needs a bit of refinement before it can be validated in CI. If anyone has desires to make such a test starting with RGW, I can pull in someone who can help with info about setting up a test AD server.

**Comment by jhoblitt** [comment]
> I'd like to note a general desire/intent to create an Active Directory (which includes LDAP) test that we could validate LDAP support in NFS and RGW. NFS still needs a bit of refinement before it can be validated in CI. If anyone has desires to make such a test starting with RGW, I can pull in someone who can help with info about setting up a test AD server.

The integration tests in this PR use openldap as the helm chart for it basically "just worked".  It has been long enough since I worked on this that my memory is fuzzy. I believe that I did try to get freeipa to go first and hit some sort of problem (which I've forgotten) and I decided to try openldap before wrestling with it. I am an active freeipa user but I deploy it via puppet.  It looks like there have been a couple of chart releases since I worked on the testing.  My guess is it probably wouldn't be that difficult to replace openldap with freeipa[+samba] for commonality between rgw and nfs tests. 

https://artifacthub.io/packages/helm/redhat-cop/ipa
https://artifacthub.io/packages/helm/improwised/freeipa

**Comment by BlaineEXE** [comment]
> The integration tests in this PR use openldap as the helm chart for it basically "just worked". It has been long enough since I worked on this that my memory is fuzzy. I believe that I did try to get freeipa to go first and hit some sort of problem (which I've forgotten) and I decided to try openldap before wrestling with it. I am an active freeipa user but I deploy it via puppet. It looks like there have been a couple of chart releases since I worked on the testing. My guess is it probably wouldn't be that difficult to replace openldap with freeipa[+samba] for commonality between rgw and nfs tests.

No worries! And thanks for the reminder about the test in place in this PR. Do you foresee any reason why a test using AD instead wouldn't properly test the LDAP features you need in RGW?

**Comment by jhoblitt** [comment]
> Do you foresee any reason why a test using AD instead wouldn't properly test the LDAP features you need in RGW?

No. Assuming the backend is 389ds/freeipa, it would be closer to my production environment.

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/types.go:1406]: In the design, we talked about including the methodology we've been using in the CephNFS using `ConfigFileVolumeSource` so that users can use Secrets, hostPath, ConfigMap, or whatever other source seems appropriate to them. It would be easiest to follow the pattern for `KerberosKeytabFile` here.

In the design, this would be Credential (`credential`) and then the ConfigFileVolumeSource (`volumeSource`) would be the element of that present. And omitting `credential` is possible if the user wishes to inject their own credential via other means.

File: .github/workflows/rgw-ldap-test/action.yml
Comment (BlaineEXE) [inline .github/workflows/rgw-ldap-test/action.yml:63]: Not needed since logs are collected by the canary test calling this.

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/types.go:1401]:
```suggestion
	DNAttribute string `json:"dnattr,omitempty"`
```

Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/types.go:1401]: The design doc has `security.ldap.config` for these params (and the last one).

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

**Pull request closed without merging**
